File: Algorithm_374_Guess Number Higher or Lower.cpp
#include <iostream>
using namespace std;
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
int guess(int num);

class Solution {
public:
    int guessNumber(int n) {
        int low=1,high=n;
        if (!guess(n)) return n;
        int mid;
        while (true)
        {
            mid=(high-low)/2+low;
            if (guess(mid)==0) return mid;
            else if (guess(mid)==1) low=mid+1;
            else high=mid;
        }
    }
};

File: Algorithm_374_Guess Number Higher or Lower.java
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * int guess(int num);
 */

 // 方法：直接二分查找，两边闭区间

public class Solution extends GuessGame {
    public int guessNumber(int n) {
        int left = 1, right = n;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int guess = guess(mid);
            if (guess == 0) {
                return mid;
            } else if (guess == 1) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }
}